Enum property with custom int discriminants

Bogay opened the pull request:
Fix #764

Bromeon commented on gdnative-core/src/nativescript/init/property/hint.rs:
```diff
@@ -116,30 +119,31 @@ where
-        }
-
-        s.into()
+        let keys: Vec<String> = self.values.keys().cloned().collect();
+        keys.join(",").into()
```
It looks like you're not using the values (i.e. associated numbers), are you?

Great how you reduced the code to be much more idiomatic 🙂

Bogay replied:
> It looks like you're not using the values (i.e. associated numbers), are you?

I didn't aware of that before, so I read the code about export a property in gdnative, and found this line at  [`godot-headers`](https://github.com/godotengine/godot-headers), it seems that I can't export information more than enum names?  If so, maybe I should find a workaround to export enum name and its corresponding value?

https://github.com/godotengine/godot-headers/blob/4bae481fc305639a076a69485bed37679149bd32/nativescript/godot_nativescript.h#L56

Bromeon replied:
I'll reply to the whole issue, as this affects how we continue.

Bromeon commented on gdnative-core/src/nativescript/init/property/hint.rs:
```diff
@@ -116,30 +119,31 @@ where
+    values: HashMap<String, i32>,
+}
+
+impl From<HashMap<String, i32>> for EnumHint {
+    #[inline]
+    fn from(values: HashMap<String, i32>) -> Self {
+        EnumHint { values }
+    }
 }
```
I would rather add a method in `EnumHint`, e.g. the suggested `with_numbers()` taking tuples. It allows to create `EnumHint` from a literal syntax and is consistent with the existing `new()` constructor -- there's no `From` implementation for `Vec<String>` either.

Construction could look as follows:
```rs
EnumHint::with_numbers(vec![
    ("one".into(), 1),
    ("two".into(), 2),
    ("seven".into(), 7),
])
```

Bromeon commented on gdnative-core/src/nativescript/init/property/hint.rs:
```diff
@@ -116,30 +119,31 @@ where
 #[derive(Clone, Eq, PartialEq, Debug, Default)]
 pub struct EnumHint {
-    values: Vec<String>,
+    values: HashMap<String, i32>,
```
My edit on the post came a bit late, maybe you missed it:
> Unless we need properties of a map somewhere, I would keep things simple and stay with `Vec`.

It's probably better for such little-used things, as `Vec` is much cheaper (performance- and memory-wise).

Bromeon reviewed (changes requested):
Thank you for your contribution! I added a few comments.

Once your version is final, you can run `cargo fmt` to auto-format everything. 🧹

Bromeon commented:
Your discovery means that Godot doesn't use numbers in the hint at all. Which probably makes sense, since it uses enum names (strings) to abstract from these very numbers.

But it also means we cannot implement this PR as it is now -- there's no point having private numeric values inside `EnumHint`, when we don't use them anywhere.

---

We should probably take a step back and ask what problem we exactly want to solve. In the issue, we had this GDScript code:
```gdscript
enum Dir { Top = -1, Bottom = 1 }
export(Dir) var dir = Dir.Bottom
```
but this does _three_ things:
* declares an `enum Dir` -- not a real type in GDScript 3, but syntax sugar for mapping named constants to numbers (if named, this is a dictionary)
* exports a **number** (print `dir` to verify that)
* associate a `GODOT_PROPERTY_HINT_ENUM` to this export, so the user has a nice drop-down in the editor

In other words, using the [enum definition](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_basics.html#enums) and [export syntax](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_exports.html), the previous code is equivalent to this:
```gdscript
const Dir: Dictionary = {
    "Top": -1,
    "Bottom": 1,
}

export(int, "Top", "Bottom") var dir = Dir["Bottom"]
```
By the way, you can verify this easily:
```gdscript
print("Selected value: ", dir)                     # 1
print("Enum: ", Dir)                               # {Bottom:1, Top:-1}
print("Is Dict: ", typeof(Dir) == TYPE_DICTIONARY) # true
```

----

In other words, the desired functionality is already today possible with godot-rust, by exporting an int and using a list of strings as enum hint.

What might add value however is if this could be done in a more convenient way -- maybe by allowing to use a Rust `enum`?

Bogay commented:
Wow, that's more complicated than my excepted before. I need more time to learn the usage of hint and export to complete this.

Bromeon commented:
Sure, don't feel forced to address this now, I think we both underestimated it 😉

Maybe I can copy the important findings to the actual issue. So maybe before doing more implementation work, let's figure out if there's something to be done at all, and if so, what.

Bogay committed:
feat(EnumHint): store optional mapping value information

Bogay committed:
feat(derive): add `ExportEnum`

currently the implementaion is dirty. I will refactor it after verifying it can work.

Bogay committed:
test(EnumHint): test g`to_godot_hint_string`

Bogay committed:
fix(ExportEnum): handle non-enum input

Bogay committed:
test(ExportEnum): correctly deny invalid input

Bogay committed:
feat(ExportEnum): correctly return error for `from_variant`

Bogay committed:
test(ExportEnum): to/from variant

Bogay committed:
docs(ExportEnum): add macro doc

Bogay commented:
@Bromeon
I think this PR is almost ready to review. You can use [my test project](https://github.com/Bogay/export-enum-demo) to quickly verify whether it works in Godot editor.
Just build the lib, and open the project to see the `dir` property of `Move` node inside `Move.tscn`.
![圖片](https://user-images.githubusercontent.com/16890663/183258022-8dc60a0c-929e-45e0-b438-d342cfbcdc5a.png)
I will rebase these changes into 2 ~ 3 commits after it's ready to merge.

BTW, derive `ExportEnum` can add editor support to the enum, but it can't be directly access from GDScript side (e.g. call `Dir::Up` in GDScript). Should I implement it in this PR? (If so, I might need to discuss about the design)

chitoyuu commented on gdnative-core/src/export/property/hint.rs:
```diff
@@ -116,30 +116,36 @@ where
-        }
-
-        s.into()
+        self.values
+            .iter()
+            .map(|(key, val)| match val {
+                Some(val) => format!("{}:{}", key, val),
+                None => key.clone(),
+            })
+            .collect::<Vec<_>>()
+            .join(",")
+            .into()
```
The original code is iterative for a reason. While `format!`, `collect` and `join` are indeed convenient and pretty, they perform many allocations under the hood. It's often said that premature optimization is the root of all evil, but that does not mean we need to needlessly slow things down just because.

Bogay replied:
Done

chitoyuu commented on gdnative-derive/src/export_enum.rs:
```diff
@@ -0,0 +1,152 @@
+    use super::*;
+
+    #[test]
+    fn deny_non_unit_variant() -> syn::Result<()> {
```
Tests for proc-macros are better modeled as [UI tests](https://github.com/godot-rust/gdnative/tree/master/gdnative/tests/ui), as they not only allow us to check if the behavior is correct (with a real compiler, too), but also whether the error messages we show to users are up to par.

Bogay replied:
Done

chitoyuu commented on gdnative-derive/src/lib.rs:
````diff
@@ -482,6 +483,61 @@ pub fn derive_from_varargs(input: TokenStream) -> TokenStream {
+///   f1: i32
+/// }
+/// ```
+#[proc_macro_derive(ExportEnum)]
+pub fn derive_export_enum(input: TokenStream) -> TokenStream {
````
I'm not sure if this is the best way we can present this in the API. On one hand, naming the macro `Export` would be too general when it only really works on primitive enums. On the other hand, users are very unlikely to think of typing `derive(ExportEnum)` by themselves when the compiler is telling them that they're missing `Export`.

chitoyuu reviewed (commented):
Reminder that as of #964, there is now a more comprehensive solution for `ToVariant` and `FromVariant` of numeric enums, integrated into the ordinary derive macros, with `#[variant(enum = "repr")]`. Please see if you can fit your work into the existing framework, so that we don't have to maintain two separate code paths for the same thing, one being a subset of another.

Bogay committed:
feat: rename ExportEnum to Export

Bogay committed:
test(Export): add UI test

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdnative/pr-775

Bogay committed:
test(Export): derive ToVariant in doctest

Bogay committed:
test: derive repr for enum

Bogay committed:
feat(Export): report compile error for all non-fieldless enum variants

Bogay committed:
test(Export): add UI tests

Bogay committed:
docs(Export): fix example code

the repr part should be added to change the serialization behavior.

chitoyuu commented on gdnative-core/src/export/property/hint.rs:
````diff
@@ -116,12 +116,22 @@ where
 /// ```
 #[derive(Clone, Eq, PartialEq, Debug, Default)]
 pub struct EnumHint {
-    values: Vec<String>,
+    values: Vec<(String, Option<i64>)>,
 }
 
 impl EnumHint {
     #[inline]
     pub fn new(values: Vec<String>) -> Self {
+        let values = values.into_iter().map(|v| (v, None)).collect();
+        EnumHint { values }
+    }
+
+    #[inline]
+    pub fn with_numbers(values: Vec<(String, i64)>) -> Self {
+        let values = values
+            .into_iter()
+            .map(|(key, val)| (key, Some(val)))
+            .collect();
         EnumHint { values }
     }
````
Is the "partially numbered" format ever useful to represent? We should expose a constructor for that format if Godot does indeed recognize it. Otherwise, I'd prefer a type definition that makes the situation clearer for the future, e.g. the following:

```rust
pub struct EnumHint {
    kind: EnumHintKind,
}

enum EnumHintKind {
    Implicit(Vec<String>),
    Explicit(Vec<(String, i64)>),
}
```

chitoyuu replied:
I see. I guess that's a possibility, although even better I suppose would be to accept a `struct EnumHintEntry { name: String, value: Option<i64> }` to which further fields can be added in the future in case it ever becomes necessary for any reason. Whether it should be made `#[non_exhaustive]` with public fields or private with `with_*` constructors I'm not 100% certain, but personally I lean toward the latter.

chitoyuu commented on gdnative-core/src/export/property/hint.rs:
```diff
@@ -130,13 +140,22 @@ impl EnumHint {
         let mut s = String::new();
 
         let mut iter = self.values.iter();
+        let write_item = |s: &mut String, item: &(String, Option<i64>)| match item {
+            (key, Some(val)) => {
+                write!(s, "{key}:{val}")
+            }
+            (key, None) => {
+                write!(s, "{key}")
+            }
+        };
```
If the whole signature is going to be spelled out anyway it should really be a `fn` instead of a closure.

chitoyuu commented on gdnative-derive/src/export.rs:
```diff
@@ -0,0 +1,65 @@
+        .iter()
+        .map(|variant| {
+            let key = &variant.ident;
+            let val = quote! { #enum_ty::#key as i64 };
```
It seems like this assumes that the `ToVariant`/`FromVariant` implementations are derived using the official macros with `#[variant(enum = "repr")]`, which isn't necessarily the case. It'd be best if the macro additionally checked for the aforementioned attribute and returned an error if there isn't one, making sure that this assumption holds.

Bogay replied:
Do you mean to check the macro (`#[variant(enum = "repr")]`) is present on the enum, or check the variant conversion is correct? e.g. `#enum_ty::#key.to_variant()` should be integer? which allows users impl their own to/from variant.

chitoyuu replied:
Getting the value from `to_variant` would be fine as well, but that produces a runtime error, which isn't ideal. It would be useful to have some sort of annotation be the trigger of this mode of the macro anyway, since there are further kinds of `Export` implementations that might be possible to derive (Newtypes maybe?) and thus I'd rather have stricter initial conditions to avoid possible breaking changes.

Bogay replied:
So here are 2 ways to check conversion:

1. check if there is a `#[variant(enum = "repr")]` attribute attached to the derived enum.
2. check if the conversion is correct within `export_info` at runtime, maybe triggered by an attribute e.g. `#[export(runtime_check)]`? I'm not sure if that's an appropriate name.

chitoyuu replied:
What I meant by "this mode" is the mode of implementing `Export` for enums with all unit variants according to its native representation, so "checking if there is a `#[variant(enum = "repr")]` attribute attached to the derived enum" would be the applicable option out of the two.

The alternative would be checking for something like `#[export(kind = "enum")]`.

Bogay replied:
> The alternative would be checking for something like `#[export(kind = "enum")]`.

I am not sure if I understand, do you mean to use the `kind = ...` to toggle how this macro generates mapping?

Bogay replied:
I've updated to a new version, is the structure what you expected? Currently the parser part is terrible, maybe `ItemAttrBuilder` can be refactored and then used for other macros that need to parse attributes?

https://github.com/godot-rust/gdnative/blob/master/gdnative-derive/src/variant/attr/item.rs#L22

chitoyuu replied:
Yes, this is roughly what I expected. The "parser" looks fine to me actually. Maybe you can use control structures like `let $pat = $expr else { $diverge };` to flatten the `if-else` hierarchies a bit, but in general I don't think the lack of a common abstraction here is a huge problem. Attribute syntax is very free-form and from my experience there are sometimes cases where the syntax most intuitive to humans turn out to require one-off, not-easily-abstracted logic for machines.

Also, there now exists middleware implementing the parsing abstraction like [`darling`](https://crates.io/crates/darling) that wasn't really a thing back when we started. Depending on the circumstances, it might be better to depend on a library like it than to roll our own. This requires quite a bit of deliberation which I don't think is appropriate to settle within this PR.

chitoyuu commented on gdnative-derive/src/export.rs:
```diff
@@ -0,0 +1,65 @@
+
+    let impl_block = quote! {
+        impl ::gdnative::export::Export for #enum_ty {
+            type Hint = ::gdnative::export::hint::IntHint<i64>;
+            #[inline]
+            fn export_info(hint: Option<Self::Hint>) -> ::gdnative::export::ExportInfo {
+                if let Some(hint) = hint {
+                    return hint.export_info();
```
What's the purpose of this bypass path here? I would assume there's not much point in deriving the enum hint but not using it on the user's part. Do you have any examples?

If a type isn't intended to be hinted further, a "never" type should be used for the `type Hint`. More information on this here: https://godot-rust.github.io/docs/gdnative/master/gdnative/export/trait.Export.html#associatedtype.Hint

Bogay replied:
Since `NoHint` is not public, we can't set `type Hint` to it. I think we should make it public? Creating an inhabitable type inside a macro does not seem like a good idea.

chitoyuu replied:
No. `NoHint` is `#[doc(hidden)] pub` so it can be used in a macro. Remember to use [`proc_macro_crate`](https://docs.rs/proc-macro-crate/latest/proc_macro_crate/) to get the path to the `gdnative` / `gdnative-core` crates. You can find examples from other existing macros.

Bogay replied:
> I don't have any particular problem with creating an inhabitable type within a macro. What do you think is bad about it?

I am not sure how to ensure that a generated type name is unique. Is using input ident as a prefix enough? or should I allow users to override the generated hint name?

chitoyuu replied:
You don't need a unique name for the type. Just group all the items into a anonymous const scope:

```rust
const _: () = {
    enum __NoHint {}
    impl Export for SomeType {
        type Hint = __NoHint;
        // ...
    }
};
```

chitoyuu commented on test/src/test_export_enum.rs:
```diff
@@ -0,0 +1,30 @@
+    ok
+}
+
+crate::godot_itest!(test_from_variant {
+    assert_eq!(Dir::from_variant(&1_i32.to_variant()), Ok(Dir::Up));
+    assert_eq!(Dir::from_variant(&(-1_i32).to_variant()), Ok(Dir::Down));
+    // 42 isn't mapped to any variant of `Dir`
+    assert!(Dir::from_variant(&42_i32.to_variant()).is_err());
+});
+
+crate::godot_itest!(test_to_variant {
+    assert_eq!(Dir::Up.to_variant(), 1_i32.to_variant());
+    assert_eq!(Dir::Down.to_variant(), (-1_i32).to_variant());
+});
```
These tests are unnecessary. A more comprehensive version already exists at https://github.com/godot-rust/gdnative/blob/6dc972190f6cefb15eef6588c81a015983718df5/test/src/test_derive.rs#L173

Bogay replied:
Got it. I will remove them. Do you think there is any other testcases should be added?  
Whether Godot rendering the property correctly seems to be a hard thing to check, I have no idea how to do that in integration tests.

chitoyuu replied:
Yeah, I don't think that's particularly easy to automatically test. Technically I guess you can inspect the inspector from the editor tests, but that's going to be very fragile given the scene structure of the editor is I assume considered an implementation detail from Godot's perspective. I think the tests are fine as is.

chitoyuu reviewed (commented):
Thanks for all the effort! The general design looks good to me, although there are still a number of minor concerns on the implementation details. See below.

I understand that it has been a long journey by now, so if you don't see yourself working any further on this PR, feel free to let me know if you'd like me to carry out any of the changes on your behalf.

bors try

Bogay committed:
feat(EnumHint): add EnumHintEntry to store entry info

Bogay committed:
test: remove duplicated tests

chitoyuu commented on gdnative-core/src/export/property/hint.rs:
```diff
@@ -116,12 +116,22 @@ where
+        let values = values.into_iter().map(EnumHintEntry::new).collect();
+        EnumHint { values }
+    }
+
+    #[inline]
+    pub fn with_values(values: Vec<(String, i64)>) -> Self {
+        let values = values
+            .into_iter()
+            .map(|(key, value)| EnumHintEntry::with_value(key, value))
+            .collect();
         EnumHint { values }
     }
```
Would it be better to simply expose the `EnumHintEntry` interface directly here instead? `fn with_entries(values: Vec<EnumHintEntry>)`

chitoyuu commented on gdnative-core/src/export/property/hint.rs:
```diff
@@ -136,13 +146,46 @@ impl EnumHint {
         }
 
         for rest in iter {
-            write!(s, ",{rest}").unwrap();
+            write!(s, ",").unwrap();
+            write!(s, "{rest}").unwrap();
         }
```
Now that there's no longer a `write_item` function necessary, this can be reverted completely to what it was.

chitoyuu reviewed (commented):
A few further things. Now all that's left should be the attribute trigger discussed in the open thread above.

bors try

Bogay committed:
feat(Export): add #[export(...)] attribute

Bogay committed:
test(ui): use glob to match tests for #[derive(Export)]

Bogay reviewed (commented):
This PR should be almost done. If the code is OK, I would update the docs.

chitoyuu reviewed (approved):
Looks good to me now. The added UI tests are nice too! Please do whatever else you meant to do and squash the remaining commits. It should be good to merge after that.

Thanks for following through this entire time!

bors try

Bogay commented:
> Looks good to me now. The added UI tests are nice too! Please do whatever else you meant to do and squash the remaining commits. It should be good to merge after that.
> 
> Thanks for following through this entire time!
> 
> bors try

@chitoyuu Sorry, I found that if `EnumHint` is used for string hint, it shouldn't have explicit mapping, or the conversion might be broken. While that's a misuse, `StringHint::Enum` with mapping is a valid object in Rust. Perhaps we should implement two types to support these two scenarios? I haven't tested for `StringHint::File` and `StringHint::GlobalFile`, but my guess is that they should be similar to `StringHint::Enum`.

chitoyuu commented:
Interesting. I think there are two options here:

1. Have `ClassBuilder` complain about it when a user tries to register a property with `StringHint` and explicit mappings. The function can strip (ignore) the values, ignore the whole property (leaving library somewhat usable but likely not behaving as expected), or just panic during init (rendering the resulting library completely unusable).
2. Revert `EnumHint` and introduce a new type (`EnumEntryint`?) for just the `IntHint` version, as a new variant.

I think option 1 with the soft warning would be the most friendly to users and developers, even though it doesn't adhere strictly to the correctness-by-construction principle. I think this is a case where the input is "correct enough" to be acceptable, and thus it's not worth the increased complexity and duplicated code paths to reject at compile time.

Which one of the options do you think is best, or maybe something else?

Bogay commented:
> Which one of the options do you think is best, or maybe something else?

Currently I prefer to warn the user when explicit mappings are given to `StringHint`, but I am not sure where the check should be inserted. The only option I see is `StringHint::export_info` when extracting the hint string.

https://github.com/godot-rust/gdnative/blob/6dc972190f6cefb15eef6588c81a015983718df5/gdnative-core/src/export/property/hint.rs#L363-L367

Maybe something like this? But that looks like a hack...

```rs
        let hint_string = match self {
            SH::Enum(e) | SH::File(e) | SH::GlobalFile(e) => {
                if e.has_value() {
                    godot_warn!("StringHint should not be used with EnumHint with mappings");
                }
                e.to_godot_hint_string()
            }
            SH::Placeholder { placeholder } => placeholder.into(),
            _ => GodotString::new(),
        };
```

Other options (strip mappings, ignore property, or separate types) all seem to require some additional API changes and more complexity. If we allow breaking changes, I would prefer to introduce a separate type for enum hint with mapping. However, I think it is better to maintain backwards compatibility in this case.

chitoyuu commented:
I suppose `StringHint::export_info` isn't the best place to emit a warning from, but it should be acceptable. Not pretty, but acceptable. You *can* frame it as sanitation during conversion to make the idea more palatable if you want, I guess. The function does take `self` by ownership so I think you should be able to implement value stripping without making a breaking change, no?

> If we allow breaking changes, I would prefer to introduce a separate type for enum hint with mapping. 

That shouldn't be a more breaking a change than what's already in tree since all the public enums now have `#[non_exhaustive]` attached to them:

```rust
enum IntHint {
    Enum(EnumHint), // old/current type without value
    EnumBikeshed(EnumBikeshedHint), // new type with EnumHintEntry
    // ...
}

enum StringHint {
    // Keep as is
}
```

> However, I think it is better to maintain backwards compatibility in this case.

I agree.

Bogay commented:
> I suppose `StringHint::export_info` isn't the best place to emit a warning from, but it should be acceptable. Not pretty, but acceptable. You _can_ frame it as sanitation during conversion to make the idea more palatable if you want, I guess.

Do you have a suggestion? I'm sorry that I don't know how to move this check to somewhere like `PropertyBuilder::done` (which might be more natural?), because it only gets the `Hint` object, which I can't access the `EnumHint` that may be in it.

> The function does take `self` by ownership so I think you should be able to implement value stripping without making a breaking change.

Do you mean something like this?

```rs
        let hint_string = match self {
            SH::Enum(e) | SH::File(e) | SH::GlobalFile(e) => {
                if e.has_value() {
                    godot_warn!("StringHint should not be used with EnumHint with mappings");
                }
                // strip all values stored inside `EnumHint`
                e.strip();
                e.to_godot_hint_string()
            }
            SH::Placeholder { placeholder } => placeholder.into(),
            _ => GodotString::new(),
        };
```

I am not sure if it would be better to implement some general purpose APIs in this case. What do you think?

```rs
        let hint_string = match self {
            SH::Enum(e) | SH::File(e) | SH::GlobalFile(e) => {
                if e.iter().any(|ent| ent.value.is_some()) {
                    godot_warn!("StringHint should not be used with EnumHint with mappings");
                }

                // strip all values stored inside `EnumHint`
                for ent in e.iter_mut() {
                    ent.value = None;
                }

                e.to_godot_hint_string()
            }
            SH::Placeholder { placeholder } => placeholder.into(),
            _ => GodotString::new(),
        };
```

chitoyuu commented:
> Do you have a suggestion? I'm sorry that I don't know how to move this check to somewhere like `PropertyBuilder::done` (which might be more natural?), because it only gets the `Hint` object, which I can't access the `EnumHint` that may be in it.

I guess we can also add a `sanitize(&mut self, errors: &mut SanitizeErrors)` function to the `Hint` trait and give it an empty default implementation so it isn't a breaking change for external implementations. `SanitizeErrors` would be something the implementer can push `std::error::Error`s to, which internally can simply be a `String` for now. This would allow for a better separation between the two steps logically, if emitting an error message from `export_hint` somehow feels sacrosanct.

> I am not sure if it would be better to implement some general purpose APIs in this case. What do you think?

I think the second example is better. I don't sense any use cases for the functions outside this one case. If they ever become necessary we can always extract the code into functions later.
